zebra-chain/src/work/arbitrary.rs

```diff
@@ -17,3 +19,81 @@ impl Arbitrary for equihash::Solution {
 
     type Strategy = BoxedStrategy<Self>;
 }
+
+impl Arbitrary for CompactDifficulty {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (vec(any::<u8>(), 32))
+            .prop_filter_map("zero CompactDifficulty values are invalid", |v| {
+                let mut bytes = [0; 32];
+                bytes.copy_from_slice(v.as_slice());
+                if bytes == [0; 32] {
+                    return None;
+                }
+                // In the Zcash protocol, a CompactDifficulty is generated using the difficulty
+                // adjustment functions. Instead of using those functions, we make a random
+                // ExpandedDifficulty, then convert it to a CompactDifficulty.
+                ExpandedDifficulty::from_hash(&block::Hash(bytes))
+                    .to_compact()
+                    .into()
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+impl Arbitrary for ExpandedDifficulty {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        any::<CompactDifficulty>()
+            .prop_map(|d| {
+                // In the Zcash protocol, an ExpandedDifficulty is converted from a CompactDifficulty,
+                // or generated using the difficulty adjustment functions. We use the conversion in
+                // our proptest strategy.
+                d.to_expanded()
+                    .expect("arbitrary CompactDifficulty is valid")
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+impl Arbitrary for Work {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: ()) -> Self::Strategy {
+        // In the Zcash protocol, a Work is converted from an ExpandedDifficulty.
+        // But some randomised difficulties are impractically large, and will
+        // never appear in any real-world block. So we just use a random Work value.
+        (any::<u128>())
+            .prop_filter_map("zero Work values are invalid", |w| {
+                if w == 0 {
+                    None
+                } else {
+                    Work(w).into()
```

Comment:
Is this `into` just wrapping the `Work` with Some? I'd prefer to be explicit on this one

Reply:
Thanks for the style tip, I'll check every `into` in this PR, and in the difficulty module.

Reply:
Fixed in 50c917ce578e9feace3a805d45e149a395edd9d0